### ISSUE TITLE
Extend config.yaml to read locations where to search for commands

### ISSUE THREAD

**Issue by alalazo** [spack/spack#7967]
### TLDR

`config.yaml` should be extended with a key that specifies the locations where to search for commands

### Rationale

Currently every command in Spack must reside in the `lib/spack/spack/cmd` folder as a standalone module. This makes extending Spack with commands that are not covering mainstream use cases (like #5545, #6166 or #7899) quite hard. 

There are, on one side, rightful concerns over adding commands on top of the ~60 existing ones - especially if they are not fitting everybody's needs. On the other side, this makes it difficult for people to publish, maintain or share their extensions, as they need to constantly rebase or merge multiple repositories.

What I propose here is to add in `config.yaml` a key where to read a list of locations to be searched for commands. This would allow to maintain Spack extensions in a separate repository (even within the Spack organization if it makes sense).

If there's consensus around this, I wouldn't mind working on a PR. In the meanwhile I'll be waiting for feedback.

**Comment by citibeth** [comment]
Can you please expand on the problems that would happen if, say, there were 1000 commands in `lib/spack/spack/cmd`?  Maybe there are other ways to solve this problem that we should consider.

Another idea would be to make it easier for people to use Spack as a Python module, just like any other Python library.  Then they could write programs that use Spack functionality without ever putting it into Spack.

**Comment by alalazo** [comment]
> Another idea would be to make it easier for people to use Spack as a Python module, just like any other Python library. Then they could write programs that use Spack functionality without ever putting it into Spack.

That's basically the same thing I am proposing here. The idea is that after this modification no user-facing aspect will change, but we will have a hooking mechanism to extend Spack with commands from outside the main Spack repository. 

This is somehow similar in concept to what [click](http://click.pocoo.org/6/) or [py-test](https://docs.pytest.org/en/latest/) do. They have a core API that can be used, their built-in functionality is based on a plug-in mechanism, but you can also extend these packages from outside their main repo. See for instance [click-spinner](https://github.com/click-contrib/click-spinner) as a small example of this.

The implementation details may vary here, but I think we could start with something simple that works just with the current mechanism Spack supports. Only, instead of looking just into `lib/spack/spack/cmd` for commands we'll read a list of directories from the configuration file first. Whenever we'll make progress towards #1385 we could look into more pythonic approaches, if they make sense.

> Can you please expand on the problems that would happen if, say, there were 1000 commands in lib/spack/spack/cmd?

The biggest problem is that an approach like that simply doesn't scale. A thousand commands in Spack means (at least to me) that we commit to maintain them all - and we might end up with commands that partially overlap in scope and with a UI that is cluttered (all the commands will be there by default).

Another problem is that currently we are complicating the life of people that want to distribute their Spack extension to somebody. The way to do it right now is to check-out a Spack fork or cherry-pick commits from different sources and be sure to follow:

1. The different branch from which you cherry-picked
2. The "base" `spack/Spack` release / branch that you want to use

and merge them constantly if you want to be up-to-date. If we give people the opportunity to write a new command and plug it into Spack we'll just need to update different independent repos to have the same functionality.

In the long run that can also work as an incubating stage for new commands: they could be extensions made by a third party until they are mature / used enough, and then we could give people the possibility to incorporate them either as:

1. "Official" plug-in hosted in a repository of the Spack organization
2. Commands that are shipped with Spack, if it makes sense to move them into the main repo

**Comment by citibeth** [comment]
I think there are simpler ways to achieve these goals.

There are two different kinds of code that might use Spack: (1) a new Spack command, complete with command line parser, etc, and (2) an arbitrary Python program that calls code from the `spack` and `llnl` modules.  I think that users might want to do both these kinds of things.  Some things users need to do, they could probably do either way; and in that case, we would want to support them the easiest way we know how.

Here are ways we could support that, from easiest to hardest:

1. Create a `spacklib` package that downloads and installs a particular version of Spack, making the code in `lib/` (not `bin/`) it available as a Python library.  This would be pretty trivial.  However, note that it is not possible for Spack to create a VERIFIED download of itself.  (There is no reason to wait for PyPI on this one).

2. Comb over the current Spack initialization to make sure that Spack data structures (specs, etc) all work without going through the main Spack entry point.  This means probably distilling out an initialization procedure that can be run from within other Python programs, without having to parse command lines and other stuff that doesn't make sense from an API.

3. Add to the `spacklib` a special way for Spack to "install" itself as a Python library for a special version (eg: "this").  This means, basically, the install procedure does nothing, and the post-install hook generates a module that points to the Spack installation.  Thus, `spack load spack@this` and `spack activate spack@this` would be two ways to make Spack's API available to Python programs (Python2 or Python3).

4. Create a new folder `lib/spack/spack/cmd/contrib`, with the operational policy that we will merge new commands into it with minimal vetting, and will not test or maintain them.  This allows users to put custom / provisional commands in there.  To access them, Spack would have to be run with the `--contrib` flag.

PS: In practice, I believe plug-in systems are a barrier to use.  There are so many these days, and using each one effectively requires some attention.  If possible, I think it's better to have the available plug-ins already plugged in.

**Comment by alalazo** [comment]
@citibeth I think there's been some misunderstanding and the conversation is deviating from my current proposal, which is really  much simpler than what you present above. When I wrote:
> The implementation details may vary here, but I think we could start with something simple that works just with the current mechanism Spack supports

I just mean to change:

https://github.com/spack/spack/blob/4bf7ce7d9987706efdd1e91ada41637c1a00273d/lib/spack/spack/cmd/__init__.py#L93-L114

to search a list of directories instead of assuming that everything is in `lib/spack/spack/cmd`. This list should be read from `config.yaml` and Spack should by default put `lib/spack/spack/cmd` into that list. That will grant that there will be no change from user perspective.

**Comment by alalazo** [comment]
@citibeth I prefer the original title, as it reflects what I propose to change in Spack

**Comment by citibeth** [comment]
Summary
----------
I think we should just hard-code `cmd/contrib` in addition to `cmd`, NOT make it a config parameter, and have a policy of merging new commands into `cmd/contrib` with minimal review.

Details
-------
@alalazo To summarize: your proposal is to create a plug-in framework for Spack commands; and to do this by replacing a previously hard-coded path with a configuration parameter.

I believe this is a an example of the aphorism of David Wheeler:
```
All problems in computer science can be solved by another level of indirection
```
In this case, I don't think it's a good idea.  Additional configuration parameters (points of indirection) add complexity to the end user experience.  A system that gets the job done with 5 configuration parameters is better than one with 100 that gets the same job done.

Consider, for example, configurable package repositories, also controlled by configuration parameters:
http://spack.readthedocs.io/en/latest/repositories.html

Configurable repos are similar in structure to the feature proposed in this Issue, except they would probably be used more often --- since more people have custom packages they want to use, than custom commands.  And yet... to this date, it is a feature of Spack that I have never used, nor have I heard of anyone else using it.  Somebody please tell me about *any* use cases where this feature is being used.  But it still takes up a whole section of the manual (2086 words), PLUS it requires an additional Spack command.

One can get the same effect without configurable repos, by using Git and maintaining your own branch with the extra packages.  This is easier and more robust for a number of reasons:
 1. It requires knowledge of Git (a general tool), rather than specialized knowledge of an obscure Spack command and configuration.
 2. Git merges are trivial because the extra packages you are maintaining are different from the ones in standard Spack.  Incompatibilities introduced by changes in Core Spack have to be addressed either way.
 3. It's easier to tell your users "check out this branch of Spack from my fork that has everything you need," rather than "check out the standard Spack, then check out my extra repo of custom packages, then edit your Spack configuration to make one point to the other."
 4. If you are administering a site, you probably want to maintain control over your software versions anyway.  The best way to do this is to choose a specific version of Spack and use it until it no longer meets your needs --- and then leap to a new version of Spack, which will likely involve rebuilding a lot of your software.  In this way, Spack is different from typical software.

All the above reasons apply equally well to the proposal in this Issue.

Other examples of systems with plug-in frameworks including Apache, Firefox, Sphinx.  It's annoying... because you can't just install the package and have it work.  You need to install the package, then discover which plug-ins /extensions / whatnots you need for YOUR problem.  Then figure out how to install them too, which usually involves learning things you didn't want to know while wading through hundreds of config parameters.  These kinds of systems have sucked many valuable hours out of my life.

**Comment by alalazo** [comment]
> Somebody please tell me about any use cases where this feature is being used

https://github.com/HSF/hep-spack
https://github.com/epfl-scitas/spack-repo-externals

**Comment by citibeth** [comment]
Thanks!

On Sat, May 5, 2018 at 4:30 PM, Massimiliano Culpo <notifications@github.com

**Comment by tgamblin** [comment]
> And yet... to this date, it is a feature of Spack that I have never used, nor have I heard of anyone else using it. Somebody please tell me about any use cases where this feature is being used.

We use this heavily for export-controlled and other internal Spack packages.  So, AFAIK, do pretty much all the national labs.  Not just for their ECI stuff, but for internal overrides to packages that maybe aren't ready for upstreaming.  I think you don't hear about it because most people who use it are using it internally, which is the whole point.  I think it would be used more publicly if we made the fetching automatic -- right now repos have to live in the filesystem; they should just be usable via github.  Also, this is hardly a foreign concept for package managers in general. Yum, RPM, Debian, etc. have channels, Homebrew has taps, etc. etc.

Homebrew and git both also have a capability for custom commands, like @alalazo is asking for.

**Comment by tgamblin** [comment]
@alalazo: I have no problem with this -- it lets people play around with commands outside Spack.  If there's something we don't want merged to the mainline, people can still use it themselves, which IMHO is great.  It would also probably encourage people to write more commands.  It's a lot like [brew external commands](https://github.com/Homebrew/brew/blob/master/docs/External-Commands.md), which is pretty much the [same thing `git` does](http://thediscoblog.com/blog/2014/03/29/custom-git-commands-in-3-steps/).

I would avoid adopting an interface as extensive as `brew`'s, where you can do shell scripts in addition to python.  I don't see where the shell option would be all that useful given the way spack internals work, and it's another thing to support.  I think it would also be a pain to provide good help text.

This should be really easy to implement, but you need more than `get_module`.  `spack.cmd.all_commands` should probably be replaced with a function (which is good anyway, as it will avoid the `os.listdir` spack currently does on startup) and it should search the list of possible command locations.  Also, you should preserve the laziness of the current command loading mechanism -- it keeps startup time low.  I don't think that requires any special treatment, but keep it in mind.

Finally, I think there's a precedence question to be answered.  I *believe* git disallows overriding builtin commands with custom commands.  I think that's probably a good idea.
